\documentclass[../../main.tex]{subfiles}

     
\begin{document}

\section{Contig classification ~\label{sec:appendix_contig_classification}}

In order to have a better understanding of the contig graph produced by a given assembler, we wish to filter out contigs that are not of chromosomal origin. %
We compare each contig against the \emph{nr} database using Megablast (Morgulis et al., 2008), and classify a contig as chromosomal if its length is greater than 1 Mb, or is such that 80\% of the first 50 Megablast hits map to a complete bacterial genome. %
We use the same second criterion to classify whether a contig is of plasmid origin, regardless of its size. %
Remaining unclassified contigs are classified as of undefined origin. %
In addition, we flag as \emph{containment} contigs those which map (using \minimap) over at least 75\% of their length to another contig.%




\section{On whether \canu contig fragmentation can be solved using \miniasm contigs \label{sec:appendix:gap_comparison}}


To check if \miniasm contigs could possibly enable to order and fill gaps between \canu contigs, we performed an assembly using the \minimap and \miniasm pipeline using both the \canu contigs and the \miniasm contigs as input (to be clear: no reads were used as input to this assembly, only two contig sets). To allow \minimap to find shorter matches, mapping of \miniasm contigs against \canu contigs was performed with the following parameters: \texttt{-x map-pb -m 25 -n 2}. To avoid \miniasm filtering overlaps,  we ran it with the following parameters: \texttt{-1 -2 -s 1000 -c 0}.

We ran this pipeline on all datasets, and counted the number of times that a \miniasm contig overlaps with two \canu contigs. We also counted the number of contigs generated by \miniasm using the overlap created at the previous step. Results are summarized in Supplementary Table \ref{tab:appendix:gap_comparison}.


\begin{table}[!htbp]
    \centering
    \begin{tabular}{l|rr|rr}
    \hline
        & \multicolumn{2}{r|}{number of genomic contig} & number of \miniasm    & number of merged contigs \\
        &  &                                            & contigs that overlap  & contigs from  \\ 
NCTC ID & \canu &  \miniasm                             & two \canu contigs & \miniasm/\canu overlaps \\ \hline
NCTC10006	&	3	&	7	&	0	&	0	\\
NCTC10332	&	12	&	22	&	1	&	0	\\
NCTC10444	&	7	&	5	&	0	&	0	\\
NCTC10702	&	3	&	2	&	0	&	0	\\
NCTC10766	&	13	&	7	&	1	&	0	\\
NCTC10794	&	7	&	5	&	0	&	0	\\
NCTC10988	&	10	&	9	&	0	&	0	\\
NCTC11126	&	7	&	15	&	5	&	0	\\
NCTC11343	&	12	&	10	&	2	&	1	\\
NCTC11360	&	26	&	25	&	1	&	0	\\
NCTC11435	&	8	&	6	&	2	&	0	\\
NCTC11800	&	7	&	3	&	1	&	0	\\
NCTC11872	&	7	&	13	&	3	&	0	\\
NCTC12123	&	5	&	3	&	2	&	0	\\
NCTC12126	&	13	&	15	&	4	&	0	\\
NCTC12131	&	16	&	77	&	3	&	0	\\
NCTC12132	&	2	&	4	&	1	&	0	\\
NCTC12146	&	3	&	1	&	0	&	0	\\
NCTC12694	&	21	&	123	&	1	&	0	\\
NCTC12841	&	16	&	1	&	0	&	0	\\
NCTC12993	&	5	&	2	&	0	&	0	\\
NCTC12998	&	3	&	4	&	0	&	0	\\
NCTC13095	&	3	&	2	&	0	&	0	\\
NCTC13125	&	6	&	7	&	0	&	0	\\
NCTC13348	&	25	&	17	&	3	&	1	\\
NCTC13463	&	5	&	4	&	1	&	0	\\
NCTC13543	&	3	&	3	&	0	&	0	\\
NCTC4672	&	68	&	16	&	5	&	1	\\
NCTC5050	&	4	&	4	&	0	&	0	\\
NCTC5053	&	8	&	11	&	2	&	1	\\
NCTC5055	&	143	&	20	&	0	&	0	\\
NCTC7922	&	13	&	9	&	4	&	1	\\
NCTC8179	&	15	&	15	&	1	&	0	\\
NCTC8500	&	3	&	1	&	0	&	1	\\
NCTC8684	&	5	&	2	&	0	&	0	\\
NCTC9075	&	7	&	3	&	2	&	0	\\
NCTC9078	&	4	&	2	&	0	&	0	\\
NCTC9098	&	8	&	6	&	3	&	0	\\
NCTC9111	&	9	&	13	&	0	&	0	\\
NCTC9112	&	7	&	15	&	10	&	0	\\
NCTC9184	&	141	&	17	&	0	&	0	\\
NCTC9645	&	31	&	76	&	9	&	3	\\
NCTC9646	&	8	&	9	&	3	&	1	\\
NCTC9695	&	2	&	1	&	0	&	0	\\
 \hline
    \end{tabular}
    \caption{
    The pipeline described section \ref{sec:appendix:gap_comparison} found more than one overlap between \canu contigs with \miniasm contig for 24 over 45 datasets. When these overlaps are re-assembled  using \miniasm, one or more merged contigs are produced in only 8 out of 45 datasets.
    }
    \label{tab:appendix:gap_comparison}
\end{table}


\newpage

\section{Assembly summary ~\label{sec:appendix_assembly_summary}}

Tables ~\ref{tb:appendix:assembly_resume} and~\ref{tb:appendix:path_search_result} report our complete results for the 45 NCTC datasets.
\begin{table}[!htbp]
\small
\begin{tabular}{l|l|r|rrr|l|rrr|rrr}
\hline
           & & & \multicolumn{3}{c|}{NCTC contigs} & \hinge  &
                                                              \multicolumn{3}{c|}{\canu contigs} & \multicolumn{3}{c}{\miniasm contigs} \\
NCTC ID    & species     &  cov & chr &  pld & und & status & chr & pld & und & chr & pld & und \\\hline
    \csvreader[head to column names]{supplemental/knot/summary_assembly.csv}{}%
    {\\\nctcid	& \textit{\species} & \coverage & \nctccontigchr & \nctccontigpld & \nctccontigund & \hingestatus & \canucontigschr & \canucontigspld & \canucontigsund & \miniasmcontigschr & \miniasmcontigspld & \miniasmcontigsund}%
    \\\hline
\end{tabular}
\caption{Datasets from the NCTC project chosen for analysis (the last row corresponds to our simulated dataset). For each sample, the coverage (cov) is given as well as the number of contigs and their assignment; chr: number of chromosomal contigs, pld: number of plasmid contigs, und: number of other contigs. For two datasets (NCTC12841 and NCTC9646) the NCTC project does not yet provide an assembly ("Pending"). For \canu and \miniasm, a classification similar to the one of NCTC is given (see text). We reported \hinge classification; FALC: Finished assembly (lacking circularization), FA: Finished assembly, MA: Mis-assembly, MA$^*$: labeled as misassembled but actually correctly solved as 2 chromosomes, FCA: Finished circular assembly, MAF: Mis-assembly/Fragmented, FAMT: Finished assembly with multiple traversals.}
\label{tb:appendix:assembly_resume}
\end{table}

\begin{table}[!htbp]
\footnotesize

\begin{tabular}{l|cc|c|cc|cccc}
\hline
 & \multicolumn{2}{c|}{\canu} & & \multicolumn{5}{c}{Edges in the AAG} \\
NCTC ID & contigs & dead-ends & dead-ends & total AAG & theoretical & distant & \multicolumn{3}{c}{adjacency edges}\\
 & & & with adj. & & max. &  edges & total & single & multiple \\
 & & & edge      & & edges &        &       &        & \\
 \hline
     \csvreader[head to column names]{supplemental/knot/path_search_resume.csv}{}%
    {\\\NCTCID & \contigcanu & \deadend & \deadendsolved & \allpath & \theoricalpath & \discard & \good & \goodnorepeat & \goodrepeat}%
    \\\hline
\end{tabular}


\caption{Assembly graph statistics for a selection of 45 fragmented assemblies from the NCTC project.  \canu assembly graph statistics: number of contigs, number of dead-end extremities. AAG statistics: theoretical maximal number number of edges. Note that for some of the most fragmented datasets (e.g. NCTC9184), none of the contigs pass the 100 Kbp length threshold, hence the AAG is empty.} 
\label{tb:appendix:path_search_result}
\end{table}


\begin{table}[]
    \centering
    \small
    \begin{tabular}{l c}
        Mean number of& \\
        \hline
         \miniasm contigs & 5.8 \\
         \hline
         Edges in AAG & 85.1\\
         Theoretical max. edges in AAG & 94.4 \\
         Distant edges & 83.12 \\
         All adjacency edges & 1.98 \\
         Single adjacency edges & 1.51 \\
         Multiple adjacency edges & 0.46 \\
         \hline
         Dead-ends in \miniasm contigs &  11.61 \\
         Dead-ends in AAG, adjacency edges & 7.95 \\ % 1.2 en non repeat
        \hline
    \end{tabular}
    \caption{Average statistics of augmented assembly graphs using a SG built from \minimap overlaps on \miniasm contigs across the 37 NCTC datasets with two or more contigs, after size and classification filters.
    All rows are as per definitions in Section~\ref{sec:aag}. 'Theoretical max. edges': number of possible edges in each AAG. 'Dead-ends in AAG, adjacency edges': number of dead-ends in the AAG when only adjacency edges are considered, i.e. distant edges are deleted.}
    \label{tab:appendix:path_search_resume_miniasm}
\end{table}


\begin{table}[!htbp]
\footnotesize

\begin{tabular}{l|cc|c|cc|cccc}
\hline
 & \multicolumn{2}{c|}{\miniasm} & & \multicolumn{5}{c}{Edges in the AAG} \\
NCTC ID & contigs & dead-ends & dead-ends & total AAG & theoretical & distant & \multicolumn{3}{c}{adjacency edges}\\
 & & & with adj. & & max. &  edges & total & single & multiple \\
 & & & edge      & & edges &        &       &        & \\
 \hline
     \csvreader[head to column names]{supplemental/knot/path_search_resume_miniasm.csv}{}%
    {\\\NCTCID & \contigcanu & \deadend & \deadendsolved & \allpath & \theoricalpath & \discard & \good & \goodnorepeat & \goodrepeat}%
    \\\hline
\end{tabular}


\caption{Assembly graph statistics for a selection of 45 fragmented assemblies from the NCTC project. \miniasm assembly graph statistics: number of contigs, number of dead-end extremities. AAG statistics: theoretical maximal number number of edges. Note that for some of the most fragmented datasets (e.g. NCTC9184), none of the contigs pass the 100 Kbp length threshold, hence the AAG is empty. '*'  denotes dataset for which the result is not available.} 
\label{tb:appendix:path_search_result_miniasm}
\end{table}

\newpage

\section{}

\begin{figure}[!htbp]
    \centering
    \includegraphics[width=0.6\textwidth]{supplemental/knot/hamilton_path.png}
    \caption{Weights of scenarios in AAGs. Each curve correspond to the sorted
list of Hamiltonian cycles, sorted by weight. If a ground truth is known,
a diamond symbol marks the correct assembly scenario}
    \label{fig:appendix:hamilton_path}
\end{figure}


\newpage

\twocolumn
\section{Contigs length and classification}
\subsection*{Canu}
\input{supplemental/knot/table_canu.tex}
\newpage
~

\newpage


\subsection*{Miniasm}
\input{supplemental/knot/table_miniasm.tex}
\onecolumn
\newpage

~

\section{Assembly length}

In Table~\ref{tb:appendix:genomic_lengths} we report the sum of lengths of all contigs in assemblies computed by \miniasm, \canu, and the assemblies produced by NCTC.

\begin{table}[!htbp]
\centering
\small
\begin{tabular}{l|rrl}
\hline
Dataset & \canu & \miniasm & NCTC\\
\hline
NCTC10006 & 5297110 & 5586964 & 5285365 \\
NCTC10332 & 6353743 & 6510723 & 6316979 \\
NCTC10444 & 5358052 & 5575543 & 5295042 \\
NCTC10702 & 3140906 & 3113949 & 3044394 \\
NCTC10766 & 5704549 & 5981360 & 5662808 \\
NCTC10794 & 2323585 & 2220964 & 2164041 \\
NCTC10988 & 3242798 & 3352903 & 3309451 \\
NCTC11126 & 4887988 & 5192073 & 4875981 \\
NCTC11343 & 6102368 & 6167977 & 5984896 \\
NCTC11360 & 2189222 & 2060590 & 2078787 \\
NCTC11435 & 4452377 & 4647620 & 4443087 \\
NCTC11800 & 4482172 & 4760149 & 4461490 \\
NCTC11872 & 1881379 & 1962919 & 1879445 \\
NCTC12123 & 4889243 & 5041564 & 4785686 \\
NCTC12126 & 5037863 & 5243635 & 5015169 \\
NCTC12131 & 4799906 & 4967498 & 4743059 \\
NCTC12132 & 3360769 & 3520932 & 3326136 \\
NCTC12146 & 5648936 & 5930232 & 5621322 \\
NCTC12694 & 4667053 & 2787030 & 4439218 \\
NCTC12841 & 1998859 & 2025517 & * \\
NCTC12993 & 5339609 & 5613096 & 5287156 \\
NCTC12998 & 5754007 & 6052839 & 5723058 \\
NCTC13095 & 6018682 & 6239133 & 5962730 \\
NCTC13125 & 5868476 & 5966557 & 3814513 \\
NCTC13348 & 5133106 & 5242304 & 5042742 \\
NCTC13463 & 5298325 & 5481195 & 5268825 \\
NCTC13543 & 5834577 & 6012216 & 5822755 \\
NCTC4672 & 2113924 & 1980239 & 1942171 \\
NCTC5050 & 5360294 & 5747419 & 5342107 \\
NCTC5053 & 5838871 & 6003270 & 5769583 \\
NCTC5055 & 5618297 & 5578757 & 4924715 \\
NCTC7152 & 5136487 & 5429453 & 5086417 \\
NCTC7922 & 5497660 & 5572395 & 5367566 \\
NCTC8179 & 5979700 & 5913131 & 5715723 \\
NCTC8500 & 4733397 & 4886947 & 4709652 \\
NCTC8684 & 4936541 & 5023386 & 4860337 \\
NCTC9075 & 5336449 & 5539086 & 5322538 \\
NCTC9078 & 5458474 & 5609092 & 5430557 \\
NCTC9098 & 5385311 & 5641175 & 5331923 \\
NCTC9111 & 5942101 & 6011787 & 5859950 \\
NCTC9112 & 5557088 & 5907239 & 5468741 \\
NCTC9184 & 2541470 & 351158 & 679813 \\
NCTC9645 & 5930294 & 5720690 & 5852841 \\
NCTC9646 & 5874126 & 6065232 & * \\
NCTC9695 & 4792855 & 4681220 & 4738566 \\
\hline
\end{tabular}
\caption{Total length of chromosomal contigs for \canu and \miniasm assemblers, total length of all contigs for NCTC on the 45 NCTC datasets studied. "*" means that NCTC assembly is not available.}
\label{tb:appendix:genomic_lengths}
\end{table}
\newpage

\section{Detailed assembly results}

\subsection*{Supplementary {\sl T. roseus} figures} 
\begin{table}[!hb]
\centering
\begin{tabular}{lr}
\hline
Total number of reads & 11592 \\
\hline
Total length & 104,608,900bp \\
\hline
Longest reads & 46,221bp \\
Shortest reads & 4bp \\
Mean Length & 9,024bp \\
Median Length & 6,978bp \\
\hline
N10 & 364 reads \\
N50 & 2586 reads \\
N90 & 7236 reads \\
\hline
L10 & 24,917bp \\
L50 & 14,590bp \\
L90 & 4,550bp \\
\hline
\end{tabular}
\caption{Some statistics about reads produced by LongISLND for \textit{T. roseus} synthetic dataset.}
\label{tb:appendix:t_roseus_20x}
\end{table}

\input{supplemental/knot/t_roseus_tig1}
\newpage

\section{Supplementary NCTC figures}

% \subsubsection*{NCTC9078}

% \begin{figure}[!hb]
% \includegraphics[width=\textwidth]{supplemental/knot/NCTC9078bog.png}
% \caption{Canu Best Overlap Graph of NCTC9078. We see two little components (bottom left): the two plasmid contigs reported in Table \ref{tb:nctc_dataset_assemblies}. The big component corresponds to the two main chromosomal contigs (tig493 and tig15). tig492 is found as a bubble in tig493. The junction  denotes a repetition: tig14 is at this junction.}
% \label{fg:appendix:bogNCTC9078}
% \end{figure}

\subsection*{NCTC12123}

\begin{figure}[!hb]

    \centerline{\includegraphics[width=\textwidth]{supplemental/knot/NCTC12123_possible_solution-crop.pdf}}
  %(a) NCTC12123 contigs mapped against NCTC reference for orderging. Paths from \canu (top dotted %edges) and \minimap (bottom dotted edges) are shown along with their number of reads as labels. Note that shortest paths in the \minimap SG from the left extremity of tig9 are searched manually because the read automatically selected did not have outgoing edges.  
%       
%        \jsv{je me demande vraiment si c'est utile ... qu'est-ce qui, dans le rapport, permet de voir cela immédiatement ?} \rc{Bonne question. On pourrait mettre ce qui concerne les overlaps en appendix dans la legende, qu'en dis-tu? }
%
    %In the \canu assembly,  tig1 and tig2 overlap over $\approx$8Kbp. tig2 and tig9 overlap over $\approx$6Kbp but we observe an intriguing drop of coverage in reads (both raw reads and those corrected by \canu) at the location of this overlap, 
    %That could explain why Canu has been unable to join those two contigs despite evidence of a path of overlapping reads unambiguously joining them. 
    %NCBI-Blast indicates that this overlap region matches a plasmid present in this species.
%
%
        %(b)
    \caption{
        Shortest paths in AAG. Both scenarii (paths that follow edges with bold label) use the edge of weight 7.178, the only remaining ambiguity concerns the order of tig1 against the pair tig2/tig9. 'overlap' indicate than our pipeline found an overlap between the contig extremities. The left scenario has a weight of 29.379 ($22.201 + 7.178$) while the right one has a weight of 30.736 ($17.209 + 7.178 + 6.349$).\ 
     }
     \label{fg:appendix:NCTC12123_solutions}
\end{figure}

~

\subsubsection*{NCTC5050}

\begin{figure}[!hb]

%\subfloat[]{
%    \centering
        \input{supplemental/knot/NCTC5050.tex}
%        \label{fg:appendix:NCTC5050_mapped_graph}
%}

%\centering\subfloat[]{
%    \includegraphics[width=0.3\textwidth]{supplemental/knot/contig_graph_NCTC5050-crop.png}
%    \label{fig:appendix:augmented_contig_graph}
%}

\caption{
NCTC5050 contigs mapped against NCTC reference for ordering. Paths are shown along with their number of bases as labels. The NCTC assembly consists of two contigs, hence the relative order of tig1 and tig10/tig23 cannot be inferred (vertical line in the figure). We observed that a portion of tig1 is inverted with respect to the NCTC assembly, with no impact on the path analysis as this putative misassembly does not involve an extremity of the contig.
%
%(b) NCTC5050 \minimap AAG generated during assembly report generation. We have 3 contig and all contig extremities are linked together, by a path found in the \minimap SG.
}
\label{fg:appendix:NCTC5050}
\end{figure}


\begin{figure}[!htbp]
    \centering
    \includegraphics[width=\textwidth]{supplemental/knot/NCTC5050_tig10_tig23.png}
    \caption{IGV view of NCTC5050 mapping of \canu contig against NCTC contig, in junction between tig10 (first track at left) and tig23 (first track at right), tig41 are mapped on begin of tig23 in forward and reverse. The second track represent the mapping of \canu corrected read, the third track represent the raw reads. Above each this track we can observe the coverage curve and drop of this curve between the tig10 and tig23, for corrected read is around 50x coverage before junction, equal to 15x at minimal, and less than 40x after junction, this value are 90x, 25x and 40x for raw read. In addition we can observe more error in corrected read on this drop of coverage.}
    \label{fig:appendix:NCTC5050_tig10_tig23}
\end{figure}

\newpage

~

\section{YACRD: Yet Another Chimeric Read Detector}

\begin{figure}[!htbp]
    \centering
    \includegraphics[width=0.6\textwidth]{supplemental/knot/yacrd.pdf}
    \caption{YACRD (manuscript in preparation) detects chimeric regions present in the read dataset. To detect such regions, YACRD takes as input the output of an overlapper (both PAF and MHAP format are accepted). For each read in the dataset, YACRD computes positional coverage values based on the overlaps with that read. If there is a drop of coverage, the corresponding read is marked as 'chimeric'.}
    \label{fig:appendix:yacrd}
\end{figure}

\end{document}